### ISSUE TITLE
Why the model are not restored in leave-one-out?

### ISSUE THREAD

**Issue by eardic** [appliedAI-Initiative/pyDVL#200]
Hi,

As far I know, the LOO algorithm for deep learning models works as below. I could not find any model restore operations in "value" and "utility" packages of yours. The aim is to find best samples for the model (by not changing the model weights) and then train only with these samples to improve accuracy of the model. If the model is not restored, then the model will be trained len(samples) times (can cause overfitting), and model weights will be changed. Do I miss anything?

Thanks a lot.
Best regards.

```
def loo(model, train_samples, test_samples):
    reference_w = model.get_weights()    # save weights for restoring later
    model.train(train_samples)
    reference_accuracy = model.test(test_samples)    
    model.restore_weights(reference_w) # not to affect the utility of remaining samples
    values = []
    for s in train_samples:         
        subset = train_samples - {s}  # the subset without the sample s
        model.train(subset) 
        values.append(reference_accuracy  - model.test(test_samples))
        model.restore_weights(reference_w) # not to affect the utility of remaining samples
   return values
```

**Comment by mdbenito** [comment]
LOO for data valuation defines the value of a point as its marginal utility, [see the docs](https://appliedai-initiative.github.io/pyDVL/30-data-valuation.html#leave-one-out-values):

$$v_u(x_i) = u(D) − u(D \setminus \{x_i\}),$$

where $u(S)$ is the performance of the model, trained on $S$, as measured on a test set. Retraining the model `len(samples)` times is a requirement for this, but has nothing to do with overfitting. One does not care about the resulting weights of each retraining at all, nor does one use the estimates of performance in any way to estimate global model performance, like is done in LOO cross-validation. 

In your example code, first you train on the whole dataset, then evaluate on the test set: this is computing $u(D)$ (your `reference_accuracy`). Then for every training sample $x_i$ you compute $u(D \setminus \\{x_i\\})$ and subtract from the total utility $u(D)$. This is computing the marginal utility as defined above, which you store it as the value for $x_i$. However, the step `model.restore_weights` at the end of (and outside) the loop has no effect whatsoever, since you call `model.train()` right after that, effectively changing all weights.

Hope that helps!

**Comment by eardic** [comment]
Hi,

I have done some research and found out that fit method of sklearn models reset all parameters and retrains with given data, which is the same as "restore_weights". So no problem :)

Thanks.
Best regards.
